### ISSUE TITLE
Ref local to struct not initialized as its real scope

### ISSUE THREAD

**Issue by sgkoishi** [icsharpcode/ILSpy#1520]
[ILSpy 5.0 Preview 2](https://github.com/icsharpcode/ILSpy/releases/tag/v5.0-preview2) yields the following code for an assembly
```c#
ref float x; // error CS8174: A declaration of a by-reference variable must have an initializer
if (flag)
{
	x = ref value1;
	x += 1f;
}
else
{
	x = ref value2;
	x += 1f;
}
```
The ref local variable x was not declared in its corresponding scope, and cause the compile error CS8174.  
It happens when it's referencing field in a struct.

**Comment by sgkoishi** [comment]
Cannot reproduce (7f2b3a)

**Comment by sgkoishi** [comment]
[ScrPad.zip](https://github.com/icsharpcode/ILSpy/files/3268935/ScrPad.zip)
This binary reproduces the problem (with a little modification).

ILSpy's result:
```c#
	public static void Main(string[] args)
	{
		ref float field; // Compile error CS8174
		if (args.Length != 0)
		{
			field = (ref Struct.Field1);
			field += 1f;
			return;
		}
		field = (ref Struct.Field1);
		field += 1f;
		if ((double)Struct.Field1 > 1.5)
		{
			field = (ref Struct.Field1);
			field += 1f;
		}
	}
```
dnSpy's result:
```c#
	public static void Main(string[] args)
	{
		bool flag = args.Length != 0;
		if (flag)
		{
			ref float ptr = ref Program.Struct.Field1;
			ptr += 1f;
		}
		else
		{
			ref float ptr = ref Program.Struct.Field1;
			ptr += 1f;
			bool flag2 = (double)Program.Struct.Field1 > 1.5;
			if (flag2)
			{
				ptr = ref Program.Struct.Field1;
				ptr += 1f;
			}
		}
	}
```

**Comment by dgrunwald** [comment]
This is a tricky case.
The control flow decompilation is independent of the local variables. It prefers using `if () { return; }` over `if-else` to keep the indentation level small.
But without the `else` block, it's impossible to declare the variable in a way that is legal C#.

**Comment by sgkoishi** [comment]
> 
> 
> This is a tricky case.
> The control flow decompilation is independent of the local variables. It prefers using `if () { return; }` over `if-else` to keep the indentation level small.
> But without the `else` block, it's impossible to declare the variable in a way that is legal C#.

Thanks for answering. Is there any option (or flag/switch) I can change to change the preference?
I assume it's in ICSharpCode.Decompiler, but didn't find it in [FlowAnalysis](https://github.com/icsharpcode/ILSpy/tree/master/ICSharpCode.Decompiler/FlowAnalysis) or [Output](https://github.com/icsharpcode/ILSpy/tree/master/ICSharpCode.Decompiler/Output).

**Comment by sgkoishi** [comment]
Fixed #1605(df13f0ce01630c44040a61b96bebee1bca252c32)
```csharp
public static void Main(string[] args)
{
	if (args.Length != 0)
	{
		Struct.Field1 += 1f;
		return;
	}
	Struct.Field1 += 1f;
	if ((double)Struct.Field1 > 1.5)
	{
		Struct.Field1 += 1f;
	}
}
```

**Comment by sgkoishi** [comment]
[BaseMod.zip](https://github.com/icsharpcode/ILSpy/files/3469373/BaseMod.zip)
BaseMod.BaseDrawing.DrawVectorChain(...)

Result:
```csharp
if (Jump <= 0f)
{
    Jump = ((float) textures[1].Height - 2f) * scale;
}
float length = 0f;
for (int i = 0; i < chain.Length - 1; i++)
{
    length += Vector2.Distance(chain[i], chain[i + 1]);
}
Vector2 start = chain[0];
Vector2 end = chain[chain.Length - 1];
Vector2 value = end - start;
value.Normalize();
float Way = 0f;
float rotation = BaseUtility.RotationTo(chain[0], chain[1]) - 1.57f;
int num = 0;
int maxTextures = textures.Length - 2;
int num2 = 0;
ref Vector2 reference = ref chain[0];
for (; Way < length; Way += Jump)
{
```
Source:
```csharp
if (Jump <= 0f)
{ 
    Jump = (textures[1].Height - 2f) * scale; 
}
float length = 0f;	
for (int m = 0; m < chain.Length - 1; m++)
{
	length += Vector2.Distance(chain[m], chain[m + 1]);
}
Vector2 start = chain[0];
Vector2 end = chain[chain.Length - 1];
Vector2 dir = end - start;
dir.Normalize();			
float Way = 0f;
float rotation = BaseUtility.RotationTo(chain[0], chain[1]) - 1.57f;
int texID = 0;
int maxTextures = textures.Length - 2;
int currentChain = 0;
Vector2 lastV = chain[0];
while (Way < length)
{
```
